hail/src/main/scala/is/hail/expr/ir/functions/DictFunctions.scala

```diff
@@ -5,30 +5,41 @@ import is.hail.expr.types
 import is.hail.expr.types._
 
 object DictFunctions extends RegistryFunctions {
+  def contains(dict: IR, key: IR) = {
+    val n = Ref(genUID(), TInt32())
+    val i = Ref(genUID(), TInt32())
 
-  def contains(dict: IR, key: IR) =
     If(IsNA(dict),
       NA(TBoolean()),
-      ArrayLen(ToArray(dict)).cne(0) &&
-        ApplyComparisonOp(
-          EQWithNA(key.typ),
-          GetField(ArrayRef(ToArray(dict), LowerBoundOnOrderedCollection(dict, key, onKey=true)), "key"),
-          key))
-
+      Let(n.name,
+        ArrayLen(ToArray(dict)),
+        n.cne(0) && Let(i.name,
+          LowerBoundOnOrderedCollection(dict, key, onKey = true),
+          If(i.ceq(n),
+            False(),
```

Comment:
This can simplify to `i.cne(n) && ApplyComparisonOp(...)`

Reply:
It wasn't short-circuiting in the Interpreter. I can try again, but I don't think it will work.

Reply:
Ah, that's fine then.

Reply:
Weird. It works now. Will change it to the nicer format.

Reply:
Ok. I understand now. It doesn't short circuit, but if there's a value of NA in the dictionary, it will never have an insertion point equal to `n` for an element not in the dictionary/set.

Reply:
Ah, so it would fail when there are no NA keys. Yeah, the way you have it is fine.